### DISCUSSION TITLE
A $refresh equivalent for Alpine v3?

### DISCUSSION THREAD

**Discussion by carltongibson** [alpine-collective/alpine-magic-helpers#147] [Q&A] [answered]
Hi. 👋. 

I'm looking for an equivalent of `$refresh` that would work for Alpine v3. 

I can't spot a hook equivalent to the old `updateElements`, so I'm wondering if you could suggest as to how this could be done? 

No doubt I'm holding it wrong but any pointers here would be greatly appreciated. 🙂  
Thanks.

**Comment by KevinBatdorf** [comment, marked as answer]
Can you share a demo of how you want to use it?

**Reply by carltongibson** [reply]
Hi @KevinBatdorf — it was quite involved — but the essence was a nested list inside an object, itself wrapped in a `Proxy`... It wasn't updating how I expected, so I was looking for  a way to force the `$refresh`. 

> No doubt I'm holding it wrong...

I found a different way, and it's working now. I **suspect** that if you do it right, the _reactivity_ will work, but (still) wondering how one would force a refresh with Alpine V3?

I shall have a potter and see if can produce a decent example of what was happening.

**Reply by KevinBatdorf** [reply]
I've been quite unwell the past week but can take a look this weekend if I'm feeling better.

I'll have to check my notes but I think there is a way to force a refresh. I can't recall now. Maybe someone else will chime in.

**Reply by carltongibson** [reply]
Thanks @KevinBatdorf — please take care of yourself before looking at this. ❤️

**Reply by HugoDF** [reply]
It might be tricky to do this in v3 since the reactivity re-renders only elements which use a fields that has changed

**Reply by KevinBatdorf** [reply]
Feeling much better today finally! Was messing around with this a bit today and came up with this sort of experimental (ok.. hacky) version that might work for you?

https://codepen.io/KevinBatdorf/pen/eYWvmpZ?editors=1010

**Reply by SimoTod** [reply]
@kevinbatdorf That example unfortunately doesn't work with x-for, see https://codepen.io/SimoTod/pen/jOmBrBe?editors=1010

I have another solution with the same defect: https://codepen.io/SimoTod/pen/NWjdjvv

The refresh helper needs to clean up x-for's and x-if's, I think

**Reply by carltongibson** [reply]
Thanks for the super examples both of you! Very helpful for me.

**Reply by KevinBatdorf** [reply]
Hmm yeah good point @SimoTod 

I tried using `clone()` initially but had other issues before. I think it only works within the same component?

I guess if we wanted to package it up we should definitely solve all the edge cases. 

I wonder how useful a `$refresh()` helper is though with most people using `Alpine.store` anyway. I sort of just made the demo quickly for fun while chatting in Discord without much consideration

**Reply by SimoTod** [reply]
Yeah, my example work only inside the component but there's no reason why it can't do a Alpine.closestRoot of a query selector to get an external component.
I used clone() because it natively handles duplicate listeners and other undesired side effects without having to really clone the node (which would kill listener created by other libraries if there are any). Still not working with x-for unfortunately, it would be good if Alpine was exposing destroyTree but it doesn't.

**Reply by KevinBatdorf** [reply]
I suppose we could just PR it. He accepted the `initTree` PR. What would be a good use case other than refresh?

**Comment by SimoTod** [comment]
Not sure if there are any. Even the refresh itself is a bit weird because Alpine should automatically track any changes.

**Reply by KevinBatdorf** [reply]
Yeah, maybe someone will come here in the future with a decent use case, but I think generally there are better ways to solve it
